### PR TITLE
Wallet model and endpoints

### LINKED ISSUE #281: Add wallet model and endpoints

### PR DESCRIPTION
**Pull request by alpkaanusen**
- Wallet model and endpoints added.

- Finish project modified to automatically make the transaction upon project completion.

### CODE REVIEW

File: backend/user/views.py [hunk 1]
```diff
@@ -320,3 +307,43 @@ def portfolio_handler(request):
...
+            return JsonResponse({"response": False, "error": "Unauthorized"})
+    elif request.method == 'PUT':
+        if token and authentication.is_authenticated(token):
+            user_id = authentication.get_user_id(token)
+            body = json.loads(request.body.decode('utf-8'))
+            try:
+                if user_id == body['user_id']:
+                    wallet = Wallet.objects.get(user=body['user_id'])
+                    wallet.balance += body['deposit'] if 'deposit' in body else 0
+                    wallet.balance -= body['withdraw'] if 'withdraw' in body else 0
```
[the remark this answers is not in the document]
Comment (emingure) [reply backend/user/views.py]: Wallet model should do this itself, So you can pass this comment :)

File: backend/user/views.py [hunk 2]
```diff
@@ -320,3 +307,43 @@ def portfolio_handler(request):
         "response": False,
         "error": "wrong request method"
     })
+
+
+@csrf_exempt
+def wallet_handler(request):
+    token = request.META.get('HTTP_AUTHORIZATION', None)
+    if request.method == 'GET':
+        user_id = request.GET.get('user_id', '')
```
Comment (enisimsar) [inline backend/user/views.py]: Redundant parameter.

File: backend/user/views.py [hunk 3]
```diff
@@ -320,3 +307,43 @@ def portfolio_handler(request):
...
+            except Exception as e:
+                return JsonResponse({'response': False, 'error': str(e)})
+        else:
+            return JsonResponse({"response": False, "error": "Unauthorized"})
+    elif request.method == 'PUT':
+        if token and authentication.is_authenticated(token):
+            user_id = authentication.get_user_id(token)
+            body = json.loads(request.body.decode('utf-8'))
+            try:
+                if user_id == body['user_id']:
```
[the remark this answers is not in the document]
Comment (enisimsar) [reply backend/user/views.py]: Agree.

File: backend/project/views.py [hunk 4]
```diff
@@ -21,6 +22,21 @@ def modify_project(json, project):
     return project
 
 
+def make_payment(project_id):
+    project = Project.objects.get(id=project_id)
+    owner_wallet = Wallet.objects.get(user=project.owner)
+    freelancer_wallet = Wallet.objects.get(user=project.freelancer)
+    bid = Bid.objects.get(project=project.id)
```
[the remark this answers is not in the document]
Comment (alpkaanusen) [reply backend/project/views.py]: Changed

File: backend/project/views.py [hunk 5]
```diff
@@ -21,6 +22,21 @@ def modify_project(json, project):
...
+    if owner_wallet.balance < bid.offer:
+        return 0  # insufficient funds
+    else:
+        owner_wallet.balance -= bid.offer
```
[the remark this answers is not in the document]
Comment (alpkaanusen) [reply backend/project/views.py]: Changed

File: backend/project/views.py [hunk 6]
```diff
@@ -220,9 +215,13 @@ def finish_project(request):
             try:
                 project = Project.objects.get(id=body['project_id'])
                 if user_id == str(project.owner.id):
-                    project.status = 2
-                    project.save()
-                    return JsonResponse({"response": True, "project": project_json(project,user_id)})
+                    payment_status = make_payment(project.id)
```
[the remark this answers is not in the document]
Comment (alpkaanusen) [reply backend/project/views.py]: Changed

File: backend/user/views.py [hunk 7]
```diff
@@ -47,6 +48,13 @@ def modify_portfolio(json, portfolio):
     portfolio.updated_at = datetime.now()
     return portfolio
 
+def create_wallet(user_id):
```
[the remark this answers is not in the document]
Comment (alpkaanusen) [reply backend/user/views.py]: Changed

File: backend/user/views.py [hunk 8]
```diff
@@ -320,3 +307,43 @@ def portfolio_handler(request):
         "response": False,
         "error": "wrong request method"
     })
+
+
+@csrf_exempt
+def wallet_handler(request):
+    token = request.META.get('HTTP_AUTHORIZATION', None)
+    if request.method == 'GET':
```
[the remark this answers is not in the document]
Comment (alpkaanusen) [reply backend/user/views.py]: Changed

**Pull request merged**